### ISSUE TITLE
Using the Sampling Override preview feature to change metrics sampling rate

### ISSUE THREAD

**Issue by stevendick-swissre** [microsoft/ApplicationInsights-Java#1665]
### Context

I'm trying to figure out if the Sampling Override preview feature can be used to reduce sampling on metrics.

The doc [here](https://docs.microsoft.com/en-us/azure/azure-monitor/app/java-standalone-sampling-overrides#common-span-attributes) doesn't help unless I want to work on HTTP or JDBC calls.

We are using Kafka and while we get all the client-side metrics for free, we get too many of them - 272 the last time I counted. We do want a subset of them, probably about 5% - 10%.

### Desired outcome

- additional examples for metrics
- links to documentation on the data model so I can work out how to write the sampling override

**Comment by trask** [comment]
hi @stevendick-swissre!

Are these metrics coming from a 3rd party library (e.g. kafka client / spring kafka)?

Are these spring boot actuator/micrometer metrics? https://docs.microsoft.com/en-us/azure/azure-monitor/app/java-standalone-config#auto-collected-micrometer-metrics-including-spring-boot-actuator-metrics

You can control the interval at which these are captured: https://docs.microsoft.com/en-us/azure/azure-monitor/app/java-standalone-config#metric-interval

We don't have a way to capture only some of these metrics, but something like this sounds like it could be a good idea:

```
{
  "instrumentation": {
    "micrometer": {
      "include": [
        "metric1",
        "metric2",
        "metric3"
      ]
    }
  }
}
```

If you don't mind, could you dump the full list of metrics that are being captured currently, and an approximate list of what you'd like to be captured? I think that would help us visualize this feature better (and what configuration would look like).

Thanks!

**Comment by stevendick-swissre** [comment]
Hi

These are the metrics coming from the Kafka client libraries (the regular Kafka client lib + the streaming lib), not from Micrometer.

We've already discovered and are using the new preview feature to change the metric interval rate - this helps.

In a first cut, we're mainly interested in monitoring lag, so we'd like to reduce the metrics to:
- kafka_consumer_fetch_manager_fetch_latency_avg
- kafka_consumer_fetch_manager_fetch_latency_max
- kafka_consumer_fetch_manager_records_lag_max
- kafka_producer_request_latency_avg
- kafka_producer_request_latency_max

Here are the metrics I see in App Insights for the Kafka client libs (281 of them):
- kafka_admin_client_connection_close_rate
- kafka_admin_client_connection_close_total
- kafka_admin_client_connection_count
- kafka_admin_client_connection_creation_rate
- kafka_admin_client_connection_creation_total
- kafka_admin_client_connections
- kafka_admin_client_failed_authentication_rate
- kafka_admin_client_failed_authentication_total
- kafka_admin_client_failed_reauthentication_rate
- kafka_admin_client_failed_reauthentication_total
- kafka_admin_client_incoming_byte_rate
- kafka_admin_client_incoming_byte_total
- kafka_admin_client_io_ratio
- kafka_admin_client_io_time_ns_avg
- kafka_admin_client_io_wait_ratio
- kafka_admin_client_io_wait_time_ns_avg
- kafka_admin_client_io_waittime_total
- kafka_admin_client_iotime_total
- kafka_admin_client_network_io_rate
- kafka_admin_client_network_io_total
- kafka_admin_client_node_incoming_byte_rate
- kafka_admin_client_node_incoming_byte_total
- kafka_admin_client_node_outgoing_byte_rate
- kafka_admin_client_node_outgoing_byte_total
- kafka_admin_client_node_request_latency_avg
- kafka_admin_client_node_request_latency_max
- kafka_admin_client_node_request_rate
- kafka_admin_client_node_request_size_avg
- kafka_admin_client_node_request_size_max
- kafka_admin_client_node_request_total
- kafka_admin_client_node_response_rate
- kafka_admin_client_node_response_total
- kafka_admin_client_outgoing_byte_rate
- kafka_admin_client_outgoing_byte_total
- kafka_admin_client_reauthentication_latency_avg
- kafka_admin_client_reauthentication_latency_max
- kafka_admin_client_request_rate
- kafka_admin_client_request_size_avg
- kafka_admin_client_request_size_max
- kafka_admin_client_request_total
- kafka_admin_client_response_rate
- kafka_admin_client_response_total
- kafka_admin_client_select_rate
- kafka_admin_client_select_total
- kafka_admin_client_successful_authentication_no_reauth_total
- kafka_admin_client_successful_authentication_rate
- kafka_admin_client_successful_authentication_total
- kafka_admin_client_successful_reauthentication_rate
- kafka_admin_client_successful_reauthentication_total
- kafka_app_info_start_time_ms
- kafka_consumer_connection_close_rate
- kafka_consumer_connection_close_total
- kafka_consumer_connection_count
- kafka_consumer_connection_creation_rate
- kafka_consumer_connection_creation_total
- kafka_consumer_connections
- kafka_consumer_coordinator_assigned_partitions
- kafka_consumer_coordinator_commit_latency_avg
- kafka_consumer_coordinator_commit_latency_max
- kafka_consumer_coordinator_commit_rate
- kafka_consumer_coordinator_commit_total
- kafka_consumer_coordinator_failed_rebalance_rate_per_hour
- kafka_consumer_coordinator_failed_rebalance_total
- kafka_consumer_coordinator_heartbeat_rate
- kafka_consumer_coordinator_heartbeat_response_time_max
- kafka_consumer_coordinator_heartbeat_total
- kafka_consumer_coordinator_join_rate
- kafka_consumer_coordinator_join_time_avg
- kafka_consumer_coordinator_join_time_max
- kafka_consumer_coordinator_join_total
- kafka_consumer_coordinator_last_heartbeat_seconds_ago
- kafka_consumer_coordinator_last_rebalance_seconds_ago
- kafka_consumer_coordinator_partition_assigned_latency_avg
- kafka_consumer_coordinator_partition_assigned_latency_max
- kafka_consumer_coordinator_partition_lost_latency_avg
- kafka_consumer_coordinator_partition_lost_latency_max
- kafka_consumer_coordinator_partition_revoked_latency_avg
- kafka_consumer_coordinator_partition_revoked_latency_max
- kafka_consumer_coordinator_rebalance_latency_avg
- kafka_consumer_coordinator_rebalance_latency_max
- kafka_consumer_coordinator_rebalance_latency_total
- kafka_consumer_coordinator_rebalance_rate_per_hour
- kafka_consumer_coordinator_rebalance_total
- kafka_consumer_coordinator_sync_rate
- kafka_consumer_coordinator_sync_time_avg
- kafka_consumer_coordinator_sync_time_max
- kafka_consumer_coordinator_sync_total
- kafka_consumer_failed_authentication_rate
- kafka_consumer_failed_authentication_total
- kafka_consumer_failed_reauthentication_rate
- kafka_consumer_failed_reauthentication_total
- kafka_consumer_fetch_manager_bytes_consumed_rate
- kafka_consumer_fetch_manager_bytes_consumed_total
- kafka_consumer_fetch_manager_fetch_latency_avg
- kafka_consumer_fetch_manager_fetch_latency_max
- kafka_consumer_fetch_manager_fetch_rate
- kafka_consumer_fetch_manager_fetch_size_avg
- kafka_consumer_fetch_manager_fetch_size_max
- kafka_consumer_fetch_manager_fetch_throttle_time_avg
- kafka_consumer_fetch_manager_fetch_throttle_time_max
- kafka_consumer_fetch_manager_fetch_total
- kafka_consumer_fetch_manager_preferred_read_replica
- kafka_consumer_fetch_manager_records_consumed_rate
- kafka_consumer_fetch_manager_records_consumed_total
- kafka_consumer_fetch_manager_records_lag
- kafka_consumer_fetch_manager_records_lag_avg
- kafka_consumer_fetch_manager_records_lag_max
- kafka_consumer_fetch_manager_records_lead
- kafka_consumer_fetch_manager_records_lead_avg
- kafka_consumer_fetch_manager_records_lead_min
- kafka_consumer_fetch_manager_records_per_request_avg
- kafka_consumer_incoming_byte_rate
- kafka_consumer_incoming_byte_total
- kafka_consumer_io_ratio
- kafka_consumer_io_time_ns_avg
- kafka_consumer_io_wait_ratio
- kafka_consumer_io_wait_time_ns_avg
- kafka_consumer_io_waittime_total
- kafka_consumer_iotime_total
- kafka_consumer_last_poll_seconds_ago
- kafka_consumer_network_io_rate
- kafka_consumer_network_io_total
- kafka_consumer_node_incoming_byte_rate
- kafka_consumer_node_incoming_byte_total
- kafka_consumer_node_outgoing_byte_rate
- kafka_consumer_node_outgoing_byte_total
- kafka_consumer_node_request_latency_avg
- kafka_consumer_node_request_latency_max
- kafka_consumer_node_request_rate
- kafka_consumer_node_request_size_avg
- kafka_consumer_node_request_size_max
- kafka_consumer_node_request_total
- kafka_consumer_node_response_rate
- kafka_consumer_node_response_total
- kafka_consumer_outgoing_byte_rate
- kafka_consumer_outgoing_byte_total
- kafka_consumer_poll_idle_ratio_avg
- kafka_consumer_reauthentication_latency_avg
- kafka_consumer_reauthentication_latency_max
- kafka_consumer_request_rate
- kafka_consumer_request_size_avg
- kafka_consumer_request_size_max
- kafka_consumer_request_total
- kafka_consumer_response_rate
- kafka_consumer_response_total
- kafka_consumer_select_rate
- kafka_consumer_select_total
- kafka_consumer_successful_authentication_no_reauth_total
- kafka_consumer_successful_authentication_rate
- kafka_consumer_successful_authentication_total
- kafka_consumer_successful_reauthentication_rate
- kafka_consumer_successful_reauthentication_total
- kafka_consumer_time_between_poll_avg
- kafka_consumer_time_between_poll_max
- kafka_producer_batch_size_avg
- kafka_producer_batch_size_max
- kafka_producer_batch_split_rate
- kafka_producer_batch_split_total
- kafka_producer_buffer_available_bytes
- kafka_producer_buffer_exhausted_rate
- kafka_producer_buffer_exhausted_total
- kafka_producer_buffer_total_bytes
- kafka_producer_bufferpool_wait_ratio
- kafka_producer_bufferpool_wait_time_total
- kafka_producer_compression_rate_avg
- kafka_producer_connection_close_rate
- kafka_producer_connection_close_total
- kafka_producer_connection_count
- kafka_producer_connection_creation_rate
- kafka_producer_connection_creation_total
- kafka_producer_connections
- kafka_producer_failed_authentication_rate
- kafka_producer_failed_authentication_total
- kafka_producer_failed_reauthentication_rate
- kafka_producer_failed_reauthentication_total
- kafka_producer_incoming_byte_rate
- kafka_producer_incoming_byte_total
- kafka_producer_io_ratio
- kafka_producer_io_time_ns_avg
- kafka_producer_io_wait_ratio
- kafka_producer_io_wait_time_ns_avg
- kafka_producer_io_waittime_total
- kafka_producer_iotime_total
- kafka_producer_metadata_age
- kafka_producer_network_io_rate
- kafka_producer_network_io_total
- kafka_producer_node_incoming_byte_rate
- kafka_producer_node_incoming_byte_total
- kafka_producer_node_outgoing_byte_rate
- kafka_producer_node_outgoing_byte_total
- kafka_producer_node_request_latency_avg
- kafka_producer_node_request_latency_max
- kafka_producer_node_request_rate
- kafka_producer_node_request_size_avg
- kafka_producer_node_request_size_max
- kafka_producer_node_request_total
- kafka_producer_node_response_rate
- kafka_producer_node_response_total
- kafka_producer_outgoing_byte_rate
- kafka_producer_outgoing_byte_total
- kafka_producer_produce_throttle_time_avg
- kafka_producer_produce_throttle_time_max
- kafka_producer_reauthentication_latency_avg
- kafka_producer_reauthentication_latency_max
- kafka_producer_record_error_rate
- kafka_producer_record_error_total
- kafka_producer_record_queue_time_avg
- kafka_producer_record_queue_time_max
- kafka_producer_record_retry_rate
- kafka_producer_record_retry_total
- kafka_producer_record_send_rate
- kafka_producer_record_send_total
- kafka_producer_record_size_avg
- kafka_producer_record_size_max
- kafka_producer_records_per_request_avg
- kafka_producer_request_latency_avg
- kafka_producer_request_latency_max
- kafka_producer_request_rate
- kafka_producer_request_size_avg
- kafka_producer_request_size_max
- kafka_producer_request_total
- kafka_producer_requests_in_flight
- kafka_producer_response_rate
- kafka_producer_response_total
- kafka_producer_select_rate
- kafka_producer_select_total
- kafka_producer_successful_authentication_no_reauth_total
- kafka_producer_successful_authentication_rate
- kafka_producer_successful_authentication_total
- kafka_producer_successful_reauthentication_rate
- kafka_producer_successful_reauthentication_total
- kafka_producer_topic_byte_rate
- kafka_producer_topic_byte_total
- kafka_producer_topic_compression_rate
- kafka_producer_topic_record_error_rate
- kafka_producer_topic_record_error_total
- kafka_producer_topic_record_retry_rate
- kafka_producer_topic_record_retry_total
- kafka_producer_topic_record_send_rate
- kafka_producer_topic_record_send_total
- kafka_producer_waiting_threads
- kafka_stream_processor_node_process_rate
- kafka_stream_processor_node_process_total
- kafka_stream_task_commit_latency_avg
- kafka_stream_task_commit_latency_max
- kafka_stream_task_commit_rate
- kafka_stream_task_commit_total
- kafka_stream_task_dropped_records_rate
- kafka_stream_task_dropped_records_total
- kafka_stream_task_enforced_processing_rate
- kafka_stream_task_enforced_processing_total
- kafka_stream_task_process_latency_avg
- kafka_stream_task_process_latency_max
- kafka_stream_task_process_rate
- kafka_stream_task_process_total
- kafka_stream_task_punctuate_latency_avg
- kafka_stream_task_punctuate_latency_max
- kafka_stream_task_punctuate_rate
- kafka_stream_task_punctuate_total
- kafka_stream_task_record_lateness_avg
- kafka_stream_task_record_lateness_max
- kafka_stream_thread_commit_latency_avg
- kafka_stream_thread_commit_latency_max
- kafka_stream_thread_commit_rate
- kafka_stream_thread_commit_total
- kafka_stream_thread_poll_latency_avg
- kafka_stream_thread_poll_latency_max
- kafka_stream_thread_poll_rate
- kafka_stream_thread_poll_total
- kafka_stream_thread_process_latency_avg
- kafka_stream_thread_process_latency_max
- kafka_stream_thread_process_rate
- kafka_stream_thread_process_total
- kafka_stream_thread_punctuate_latency_avg
- kafka_stream_thread_punctuate_latency_max
- kafka_stream_thread_punctuate_rate
- kafka_stream_thread_punctuate_total
- kafka_stream_thread_task_closed_rate
- kafka_stream_thread_task_closed_total
- kafka_stream_thread_task_created_rate
- kafka_stream_thread_task_created_total

**Comment by trask** [comment]
thanks @stevendick-swissre!

we are targeting metric filtering preview for the 3.1.1 GA release next week

will post link to docs here once it's live

**Comment by trask** [comment]
Hey @stevendick-swissre, 

Metric filters are available now in 3.1.1, please see the docs at
https://docs.microsoft.com/en-us/azure/azure-monitor/app/java-standalone-telemetry-processors#metric-filter

We didn't support `include` because of concerns how that would work with built-in metrics and jmx metrics, whereas `exclude` seems unambiguous, "I want to exclude these specific metrics".

This doesn't make it super convenient for what you want (excluding almost all of the metrics).

But since it supports regular expressions, you _could_ cheat the system by using a carefully crafted negative lookahead expression 😉, e.g.

```
"processors": [
  {
    "type": "metric-filter",
    "exclude": {
      "matchType": "regexp",
      "metricNames": [
        "kafka_(?!consumer_fetch_manager_fetch_latency_avg$|consumer_fetch_manager_fetch_latency_max$|consumer_fetch_manager_records_lag_max$|producer_request_latency_avg$|producer_request_latency_max$).*"
      ]
    }
  }
]
```
